Tools/WinMLRunner/src/EventTraceHelper.cpp

```diff
@@ -0,0 +1,559 @@
+            if (ERROR_SUCCESS != status)
+            {
+                printf("Failed to format ETW event user data..");
+
+                return status;
+            }
+        }
+        if (commandArgs.UseGPU() && 
+            wcscmp(executionProvider.c_str(), L"CPUExecutionProvider") == 0 && 
+            !operatorName.empty())
+        {
+            wprintf(L"WARNING: CPU fallback detected for operator %s(%s), duration: %s\n", operatorName.c_str(),
+                    operatorType.c_str(), duration.c_str());
+        }
+    }
+
+    return ERROR_SUCCESS;
+}
+
+VOID WINAPI EventTraceHelper::EventRecordCallback(EVENT_RECORD* pEventRecord)
+{
+    // This is where you would get the details of ETW event
+    DWORD status = ERROR_SUCCESS;
+    PTRACE_EVENT_INFO pInfo = NULL;
+    LPWSTR pwsEventGuid = NULL;
+    PBYTE pUserData = NULL;
+    PBYTE pEndOfUserData = NULL;
+    ULONGLONG TimeStamp = 0;
+    ULONGLONG Nanoseconds = 0;
+
+    // Skips the event if it is the event trace header. Log files contain this event
+    // but real-time sessions do not. The event contains the same information as
+    // the EVENT_TRACE_LOGFILE.LogfileHeader member that you can access when you open
+    // the trace.
+
+    if (IsEqualGUID(pEventRecord->EventHeader.ProviderId, EventTraceGuid) &&
+        pEventRecord->EventHeader.EventDescriptor.Opcode == EVENT_TRACE_TYPE_INFO)
+    {
+        // Skip this event.
+    }
+    else
+    {
+        // Process the event. The pEventRecord->UserData member is a pointer to
+        // the event specific data, if it exists.
+
+        status = GetEventInformation(pEventRecord, pInfo);
+
+        if (ERROR_SUCCESS != status)
+        {
+            printf("GetEventInformation failed with %lu\n", status);
+            goto cleanup;
+        }
+
+        if (DecodingSourceTlg == pInfo->DecodingSource)
+        {
+            FormatDataForCPUFallback(pEventRecord, pInfo);
+        }
+        else // Not handling any events other than Tlg type
+        {
+            // Do nothing
+        }
+    }
+
+cleanup:
+
+    if (pInfo)
+    {
+        free(pInfo);
+    }
+}
+
+ULONG WINAPI EventTraceHelper::BufferCallback(PEVENT_TRACE_LOGFILEW pLogFile) 
+{ 
+    return TRUE; 
+}
+
+void EventTraceHelper::ProcessEventTrace(PTRACEHANDLE traceHandle)
+{
+    try
+    {
+        auto status = ProcessTrace(
+            traceHandle, 1, 0,
+            0); // this call blocks until either the session is stopped or an exception is occurred in event_callback
+        if (status != ERROR_SUCCESS && status != ERROR_CANCELLED)
+        {
+            printf("ProcessTrace failed with %lu\n", status);
+        }
+    }
+    catch (exception ex)
+    {
+        printf("Exception when processing event trace: %s\n", ex.what());
+        // catch exceptions occurred in event_callback
+    }
+}
+
+void EventTraceHelper::Start()
+{
+    int bufferSize = sizeof(EVENT_TRACE_PROPERTIES) + (sizeof(LOGSESSION_NAME) + 1) * sizeof(wchar_t);

```

Comment:
>+ (sizeof(LOGSESSION_NAME) + 1) * sizeof(wchar_t); [](http://example.com/codeflow?start=52&length=50)

It looks like the start trace function describes this weird memory layout... can you add a link explaining why this extra memory is being added to the struct:

[https://docs.microsoft.com/en-us/windows/win32/api/evntrace/ns-evntrace-event_trace_properties](https://docs.microsoft.com/en-us/windows/win32/api/evntrace/ns-evntrace-event_trace_properties)